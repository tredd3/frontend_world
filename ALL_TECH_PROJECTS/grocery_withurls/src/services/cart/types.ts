import { actions } from './shared';

export type NonCheckoutParams = { action: typeof actions.preCheckout };
export type CheckoutParams = {
  action: typeof actions.postCheckout;
  // customerAddressId is actually a string in the request. Need to check if number works.
  // If not, we'll have to cast before sending it.
  customerAddressId: number;
  deliveryDate: string;
  timeslot: '';
  isAutoSelected: 0;
};

export type MaybeCheckout = CheckoutParams | NonCheckoutParams;

export type AddCartItemRequestBase = {
  userId: number;
  storeId?: number;
  cartId: number;
  product: { skuId: number; qty: number; productType: number }[];
};

export type AddCartItemRequest = AddCartItemRequestBase & NonCheckoutParams;

export type ChangeQuantityRequest = AddCartItemRequestBase & MaybeCheckout;

export type RemoveCartItemRequest = {
  userId: number;
  storeId?: number;
  cartId: number;
  skuId: number;
  qty: 0;
  type: number;
} & MaybeCheckout;

export type MoveCartItemRequest = {
  userId: number;
  cartId: number;
  skuId: number;
  qty: number;
  productType: number;
  moveType: 'SaveForLater' | 'MoveToCart';
};

export type DeleteSavedForLaterItemRequest = {
  skuId: number;
  qty: number;
  productType: number;
  productCategoryType: 'SaveForLater';
};

export type PaymentRequestRequest = {
  Amount: number;
  CartId: number;
  MobileNo: number;
  StoreId: number;
  UserId: number;
};

export type PaymentRequestResponse = {
  SUPERMID: string;
  CLIENTID: string;
  MERCHANTNAME: string;
  SPLITTXNFLAG: string;
  CHANNEL: string;
  CURRENCY: string;
  TXNTYPE: string;
  Time: string;
  Checksum: string;
  Amount: string;
  RefNo: string;
  SplitDetails: string;
  ReturnUrl: string;
  PaymentUrl: string;
  Version: string;
  IncludePayModes: string;
  ExcludePayModes: string;
  ProductDescription: string;
  UDF1: string;
  UDF2: string;
  UDF3: string;
  UDF4: string;
  UDF5: string;
};

export type PGPostPayload = {
  merchantid: string;
  clientid: string;
  merchantname: string;
  splittxnflag: string;
  channel: string;
  'transaction.currency': string;
  'transaction.txntype': string;
  'transaction.timestamp': string;
  checksum: string;
  'transaction.amount': string;
  'transaction.extref': string;
  'transaction.splitdetails': string;
  returl: string;
  version: string;
  'preferences.includepaymodes': string;
  'preferences.excludepaymodes': string;
  productdescription: string;
  udf1: string;
  udf2: string;
  udf3: string;
  udf4: string;
  udf5: string;
  token: string;
  'subscriber.customername': string;
  'subscriber.mobilenumber': string;
  'subscriber.customerid': string;
}

export type DeliveryModesAPIRequestType = {
  StoreId: number;
  OrderValue: number;
  Pincode: number;
};

export type CheckoutRequest = {
  storeId: number;
  cartId: number;
  userId: number;
  // Not sure if the user's mobile number is needed
  // userMobile: number;
} & CheckoutParams;

export type PersistedPaymentDetails = {
  refNo: string;
  userId: number;
  storeId: number;
  cartId: number;
};

export type PaymentResponseRequest = {
  RefNo: string;
  UserId: number;
  StoreId: number;
};

export type PlaceCODOrderRequest = {
  CartId: number;
  StoreId: number;
  PaymentMethod: 1;
  MID: 0;
}
