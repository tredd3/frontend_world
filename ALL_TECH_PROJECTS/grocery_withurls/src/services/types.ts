/* eslint-disable max-classes-per-file */
import { ControlType } from './helpers/make-api-call';

export class APIError extends Error {
  code: number;

  constructor({ Control }: { Control: ControlType}) {
    super(Control.Message || 'Unexpected response from API');
    this.code = Control.MessageCode;
  }
}

export class HTTPError extends Error {
  status: number;

  constructor(response: Response) {
    super(`${response.status} - ${response.statusText}`);
    this.status = response.status;
  }
}
