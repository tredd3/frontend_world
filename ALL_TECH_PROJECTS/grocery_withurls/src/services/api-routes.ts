const baseUrl = process.env.REACT_APP_BASE_URL || '';
const baseUrlLogin = process.env.REACT_APP_BASE_URL_LOGIN || '';

export const apiRoutes = {
  getNewAccessToken: `${baseUrl}/jm/auth/oauth/v2/token`,
  getDashboard: `${baseUrl}/jiokart/v1/jiogrocery/get/dashboard`,
  viewAllDashboard: `${baseUrl}/jiokart/v1/jiogrocery/view/all/dashboard`,
  getStores: `${baseUrl}/jiokart/v1/get/stores`,
  getCart: `${baseUrl}/jiokart/v1/jiogrocery/get/cart`,
  addCartItem: `${baseUrl}/jiokart/v1/add/cart/item`,
  removeCartItem: `${baseUrl}/jiokart/v1/remove/cart/item`,
  moveCartItem: `${baseUrl}/jiokart/v1/jiogrocery/move/item`,
  getCategories: `${baseUrl}/jiokart/v1/get/categories`,
  getProductList: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  checkServiceability: `${baseUrl}/jiokart/v1/jiogrocery/pincode/check`,
  getProductSearch: `${baseUrl}/jiokart/v1/jiogrocery/product/search`,
  getSuggestions: `${baseUrl}/jiokart/v1/jiogrocery/suggestion`,
  addUser: `${baseUrl}/jiokart/v1/jiogrocery/add/user`,
  saveAddress: `${baseUrl}/jiokart/v1/jiogrocery/save/user/address`,
  getAddress: `${baseUrl}/jiokart/v1/jiogrocery/get/user/address`,
  deleteAddress: `${baseUrl}/jiokart/v1/jiogrocery/remove/user/address`,
  getProductDetails: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  getSimilarItems: `${baseUrl}/jiokart/v1/jiogrocery/store/similar/items`,
  getOrderDetails: `${baseUrl}/jiokart/v1/jiogrocery/order/details`,
  getOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/listing`,
  getDeliveryModes: `${baseUrl}/jiokart/v1/get/delivery/modes`,
  getLocation: `${baseUrl}/jiokart/v1/jiogrocery/google/places`,
  searchOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/item/search`,
  checkout: `${baseUrl}/jiokart/v1/order/checkout`,
  placeOrder: `${baseUrl}/jiokart/v1/place/order`,
  getUserDetail: `${baseUrl}/jiokart/v1/jiogrocery/get/user`,
  getWishListData: `${baseUrl}/jiokart/v1/jiogrocery/get/item/list`,
  addItemToWishList: `${baseUrl}/jiokart/v1/jiogrocery/add/item`,
  removeItemFromWishList: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`,
  sendOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/send`,
  verifyOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/verify`,
  validateuser: `${baseUrl}/cr/v1/validateuser`,
  cancelOrder: `${baseUrl}/jiokart/v1/order/cancel`,
  config: `${baseUrl}/jiokart/v1/logged/customer/configurations`,
  updateUser: `${baseUrl}/jiokart/v1/jiogrocery/update/user`,
  orderRating: `${baseUrl}/jiokart/v1/order/rating`,
  getPaymentRequest: `${baseUrl}/jiokart/v1/get/payment/request`,
  getPaymentResponse: `${baseUrl}/jiokart/v1/get/payment/response`,
  getNotifications: `${baseUrl}/jiokart/v1/jiogrocery/notification/get`,
  saveNotification: `${baseUrl}/jiokart/v1/jiogrocery/notification/save`,
  clearNotifications: `${baseUrl}/jiokart/v1/jiogrocery/clear/notification`,
  removeCartSavedItem: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`
} as const;

const endpointsWithoutAuth = [
  apiRoutes.checkServiceability,
  apiRoutes.getStores,
  apiRoutes.getProductDetails,
  apiRoutes.getSimilarItems,
  apiRoutes.getLocation,
  apiRoutes.sendOtp,
  apiRoutes.getDashboard
];

const endpointsNeedingEnvelope = [
  apiRoutes.getCategories,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.getStores,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.getPaymentRequest,
  apiRoutes.getPaymentResponse,
  apiRoutes.orderRating
];

const endpointsNeedingStoreId = [
  apiRoutes.getCart,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.placeOrder,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.saveAddress,
  apiRoutes.getAddress,
  apiRoutes.getPaymentResponse,
  apiRoutes.removeCartSavedItem
];

export const needsEnvelope = (url: string) => endpointsNeedingEnvelope.includes(url);
export const needsAuthentication = (url: string) => !endpointsWithoutAuth.includes(url);
export const needsStoreId = (url: string) => endpointsNeedingStoreId.includes(url);

export const v2ApiRoutes = [
  apiRoutes.getPaymentResponse,
  apiRoutes.getOrderHistory,
  apiRoutes.getOrderDetails,
  apiRoutes.getCart,
  apiRoutes.getDeliveryModes
];
