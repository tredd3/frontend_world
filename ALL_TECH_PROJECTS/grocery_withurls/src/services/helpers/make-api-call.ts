/* eslint-disable no-console */
import uuidV1 from 'uuid/v1';
import { needsEnvelope, needsAuthentication, v2ApiRoutes } from '../api-routes';
import { getTokens } from '../../intents';
import { HTTPError, APIError } from '../types';

export type ControlType = { Status: number; MessageCode: number; Message: string };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const prepareBody = <RequestBodyType extends Record<string, any>>(body: RequestBodyType, url: string) => {
  const requestBody = {
    Request: {
      Header: {
        DeviceId: '6cf35d17989d3fbe',
        RequestTime: new Date().getTime(),
        RequestId: uuidV1(),
        Version: v2ApiRoutes.includes(url) ? 2.0 : parseFloat(process.env.REACT_APP_API_VERSION!),
        Product: Number(process.env.REACT_APP_PRODUCT_VERSION)
      },
      Body: body || {}
    }
  };

  return needsEnvelope(url) ? { Data: requestBody } : requestBody;
};

type LogType = 'success' | 'error';
type LogInfo = {
  startTime: number;
  url: string;
  fetchOptions: any;
  response: any;
  json?: any;
  type: LogType;
};
const log = ({
  startTime, url, fetchOptions, response, json, type
}: LogInfo) => {
  if (process.env.NODE_ENV === 'development') { // Compiles this block out in non-dev builds
    const { body, ...otherOptions } = fetchOptions;
    console.groupCollapsed(
      `HTTP request: ${url} %c${Date.now() - startTime}ms %c${type}`,
      'color: grey;',
      `color: ${type === 'success' ? 'green' : 'red'}`
    );
    console.log('Request body:', JSON.stringify(JSON.parse(body), null, 2));
    if (json) console.log('Response json', JSON.stringify(json, null, 2));
    console.log('Full request:', { body, ...otherOptions });
    console.log('Full response:', response);
    console.groupEnd();
  }
};

const doesBodyContainError = ({ Control }: { Control: ControlType }) => (
  !Control || Control.Status === 0 || Control.MessageCode === 404
);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MakeApiCallOptions<RequestBodyType extends Record<string, any>> =
  Partial<Omit<RequestInit, 'body'>> & { body: RequestBodyType };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default async <RequestBodyType extends Record<string, any>, ResponseType>(
  url: string,
  { headers, body, ...otherOptions }: MakeApiCallOptions<RequestBodyType> = { body: {} as RequestBodyType }
) => {
  const tokens = await getTokens();

  const startTime = Date.now();
  const fetchOptions = {
    method: 'POST',
    headers: {
      'X-API-Key': tokens.apiKey,
      'Content-Type': 'application/json',
      'X-REQUEST-ID': uuidV1(),
      'app-name': 'RJIL_JioKart',
      ...(needsAuthentication(url) && { Authorization: `Bearer ${tokens.accessToken}` }),
      ...headers
    },
    body: JSON.stringify(prepareBody<RequestBodyType>(body, url)),
    ...otherOptions
  };
  const response = await fetch(url, fetchOptions);

  if (!response.ok) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, type: 'error'
      });
    }
    throw new HTTPError(response);
  }
  const json = await response.json();

  if (doesBodyContainError(json)) {
    if (process.env.NODE_ENV === 'development') {
      log({
        startTime, url, fetchOptions, response, json, type: 'error'
      });
    }
    throw new APIError(json);
  }

  if (process.env.NODE_ENV === 'development') {
    log({
      startTime, url, fetchOptions, response, json, type: 'success'
    });
  }

  return json.Data ? json.Data as ResponseType : json.Control as ResponseType;
};
