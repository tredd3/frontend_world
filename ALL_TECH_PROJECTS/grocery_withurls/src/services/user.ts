import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { formatDate } from '../helpers/functional';
import { pivotAPIUser } from './helpers/pivots';
import { APIUser, UserPivot } from '../types';
import { genderStringToCode } from './helpers/generic';

export const getUser = promiseMemoize(async (): Promise<UserPivot> => {
  const user = await makeApiCall<{}, APIUser>(apiRoutes.getUserDetail);
  return pivotAPIUser(user);
});

type UpdateUserArgs = {
  userId?: number;
  firstName?: string;
  lastName?: string;
  email?: string;
  gender?: UserPivot['gender'];
  dob?: Date;
  lat?: number;
  lng?: number;
  pincode?: number;
  city?: string;
  storeId?: number;
  storeName?: string;
};

export const updateUser = async ({
  firstName, lastName, email, gender, dob, lat, lng, pincode, city, storeId, storeName
}: UpdateUserArgs) => {
  const userFromApi = await getUser();
  const { userId } = userFromApi;
  const isValidStoreId = storeId !== undefined;
  const isValidStoreName = storeName !== undefined;

  await makeApiCall(apiRoutes.updateUser, {
    body: {
      userId,
      ...(firstName && { firstName }),
      ...(lastName && { lastName }),
      ...(email ? { emailId: email } : { emailId: '' }),
      ...(gender && { gender: genderStringToCode(gender) }),
      ...(dob ? { dob: formatDate(dob) } : { dob: '' }),
      preferences: {
        ...userFromApi.preferences,
        ...(pincode && { pincode }),
        ...(lat && { latitude: lat }),
        ...(lng && { longitude: lng }),
        ...(isValidStoreId && { storeId }),
        ...(city && { city }),
        ...(isValidStoreName && { storeName })
      }
    }
  });

  getUser.clear();
};

type AddUserArgs = {
  firstName: string;
  lastName: string;
};

export const addUser = ({ firstName, lastName }: AddUserArgs) => (
  makeApiCall<AddUserArgs, void>(apiRoutes.addUser, {
    body: {
      firstName,
      lastName
    }
  })
);
