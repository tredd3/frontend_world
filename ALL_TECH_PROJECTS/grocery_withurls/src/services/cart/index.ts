import promiseMemoize from 'promise-memoize';
import { apiRoutes } from '../api-routes';
import makeApiCall from '../helpers/make-api-call';
import { APICart } from '../../types/api';
import { getUser } from '../user';
import { getCart as internalGetCart } from './internal';

export * from './pre-checkout';
export * from './post-checkout';
export * from './payment';

// PS: This method is only for use in old components that haven't been changed
// to use pivot data and are passing the data as such to multiple such components.
// If it is just a single component consuming the same data, refactor the component
// to rely on pivoted data instead.
export const deprecatedGetCart = promiseMemoize(async (storeId?: number): Promise<APICart> => {
  const { userId, preferences: { storeId: defaultStoreId } } = await getUser();
  const selectedStoreId = storeId || defaultStoreId;

  return makeApiCall<{}, APICart>(apiRoutes.getCart, {
    body: {
      userId,
      ...(selectedStoreId ? { storeId: selectedStoreId } : {})
    },
    ...(selectedStoreId ? { headers: { StoreId: `${selectedStoreId}` } } : {})
  });
}, { resolve: 'json' });

export const clearGetCartMemoization = () => internalGetCart.clear();
