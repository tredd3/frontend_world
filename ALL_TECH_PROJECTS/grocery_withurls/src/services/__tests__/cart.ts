import { getCart } from '../cart';
import * as makeApiCall from '../helpers/make-api-call';
import { apiRoutes } from '../api-routes';
import mockDB from '../../mocks/db.json';
import * as UserService from '../user';
import { pivotAPIUser } from '../helpers/pivots';
import { APIUser } from '../../types';

const mockUserWithPreferredStore = {
  customerName: 'Anurag Grover',
  status: 1,
  registrationDate: 1582185826567,
  addedDate: 1582185826567,
  lastUpdatedDate: 1585829941243,
  dob: '2007-07-30',
  gender: 1,
  firstName: 'Anurag',
  lastName: 'Grover',
  // eslint-disable-next-line max-len
  deviceId: 'ZDFLRFItWHJuVjg6QVBBOTFiSEZnSTU2RWRNem93UFotUU5sWGxfcm9VeXpTZFFCeGR1cVMxTlU4\nbkFaNVJQNzR4THk0ZnRSU2M0OHdBbU9lUU1saVcxRnRnTk1GRXl3dVotdVNoMGlSU1NmYmVwMGRr\neDE2N1h6Sm1keVRRTFlFRzhrZ1VPaHJFT1BsZjlUSG1JWDVKR20=\n',
  deviceType: 'ANDROID',
  preferences: {
    latitude: 19.05516658822881,
    longitude: 73.06983741392456,
    storeId: 1005474,
    pincode: 410210,
    storeName: 'Shobhana Stores',
    city: 'Navi Mumbai'
  },
  userId: 10041542,
  emailId: 'abc@gmail.com',
  mobileNumber: 9971597788
} as APIUser;

jest.mock('../../intents', () => ({
  getTokens: () => ({
    apiKey: 'dummy-api-key',
    accessToken: 'dummy-access-token'
  })
}));

describe('cart service tests for user with no selected store in preferences', () => {
  let makeApiCallSpy: jest.SpyInstance<Promise<unknown>>;
  let getUserSpy: jest.SpyInstance<Promise<unknown>>;

  beforeAll(() => {
    getUserSpy = jest.spyOn(UserService, 'getUser').mockImplementation(async () => pivotAPIUser(mockDB.user.Data));
    makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.cart.Data);
  });

  beforeEach(() => {
    makeApiCallSpy.mockClear();
    // getCart.clear();
  });

  afterAll(() => {
    makeApiCallSpy.mockRestore();
    getUserSpy.mockRestore();
  });

  test('getCart for user with no default store', async () => {
    const cartData = await getCart();
    expect(cartData).toMatchSnapshot();

    expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
    expect(makeApiCallSpy).toHaveBeenLastCalledWith(apiRoutes.getCart, {
      body: {
        userId: mockDB.user.Data.userId
      },
      headers: {
        StoreId: '0'
      }
    });
  });

  // test('addCartItem', async () => {
  //   const oldCart = await getCart();
  //   const cartData = await addCartItem({
  //     action: 0,
  //     product: [{ skuId: 1, qty: 2, productType: 2 }]
  //   });
  //   expect(cartData).toMatchSnapshot();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(3);
  //   expect(makeApiCallSpy).toHaveBeenNthCalledWith(2, apiRoutes.addCartItem, {
  //     body: {
  //       userId: mockDB.user.Data.userId,
  //       cartId: oldCart.cartId,
  //       action: 0,
  //       product: [{ skuId: 1, qty: 2, productType: 2 }]
  //     }
  //   });
  // });

  // test('removeCartItem', async () => {
  //   const cartData = await removeCartItem({
  //     cartId: 1,
  //     type: 0
  //   });
  //   expect(cartData).toMatchSnapshot();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(2);
  //   expect(makeApiCallSpy).toHaveBeenNthCalledWith(1, apiRoutes.removeCartItem, {
  //     body: {
  //       cartId: 1,
  //       qty: 0,
  //       type: 0
  //     }
  //   });
  // });

  // test('saveCartItemForLater', async () => {
  //   const cartData = await saveCartItemForLater({
  //     cartId: 1,
  //     qty: '1',
  //     productType: 2
  //   });
  //   expect(cartData).toMatchSnapshot();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(2);
  //   expect(makeApiCallSpy).toHaveBeenNthCalledWith(1, apiRoutes.moveCartItem, {
  //     body: {
  //       cartId: 1,
  //       qty: '1',
  //       productType: 2,
  //       moveType: 'SaveForLater'
  //     }
  //   });
  // });

  // test('confirmChangedCart', async () => {
  //   await confirmChangedCart();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
  //   expect(makeApiCallSpy).toHaveBeenLastCalledWith(apiRoutes.getCart, {
  //     body: {
  //       userId: mockDB.user.Data.userId,
  //       action: 1
  //     }
  //   });
  // });
});

describe('cart service tests for users with already selected store in preferences', () => {
  let makeApiCallSpy: jest.SpyInstance<Promise<unknown>>;
  let getUserSpy: jest.SpyInstance<Promise<unknown>>;

  beforeAll(() => {
    getUserSpy = jest.spyOn(UserService, 'getUser').mockImplementation(async () => pivotAPIUser(mockUserWithPreferredStore));
    makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.cart.Data);
  });

  beforeEach(() => {
    makeApiCallSpy.mockClear();
    // getCart.clear();
  });

  afterAll(() => {
    makeApiCallSpy.mockRestore();
    getUserSpy.mockRestore();
  });

  // test('getCart for user with default store selected in preferences', async () => {
  //   const makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.cart.Data);
  //   const getUserSpy = jest.spyOn(UserService, 'getUser').mockImplementation(async () => pivotAPIUser(mockUserWithPreferredStore));

  //   const cartData = await getCart();
  //   expect(cartData).toMatchSnapshot();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
  //   expect(makeApiCallSpy).toHaveBeenLastCalledWith(apiRoutes.getCart, {
  //     body: {
  //       userId: mockUserWithPreferredStore.userId,
  //       storeId: mockUserWithPreferredStore.preferences.storeId
  //     },
  //     headers: {
  //       StoreId: `${mockUserWithPreferredStore.preferences.storeId}`
  //     }
  //   });

  //   makeApiCallSpy.mockRestore();
  //   getUserSpy.mockRestore();
  // });

  // test('getCart for user with default store selected in preferences but for passed storeId parameter in call', async () => {
  //   const makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.cart.Data);
  //   const getUserSpy = jest.spyOn(UserService, 'getUser').mockImplementation(async () => pivotAPIUser(mockUserWithPreferredStore));
  //   const mockStoreId = 12344321;

  //   const cartData = await getCart(mockStoreId);
  //   expect(cartData).toMatchSnapshot();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
  //   expect(makeApiCallSpy).toHaveBeenLastCalledWith(apiRoutes.getCart, {
  //     body: {
  //       userId: mockUserWithPreferredStore.userId,
  //       storeId: mockStoreId
  //     },
  //     headers: {
  //       StoreId: `${mockStoreId}`
  //     }
  //   });

  //   makeApiCallSpy.mockRestore();
  //   getUserSpy.mockRestore();
  // });

  // test('addCartItem', async () => {
  //   const makeApiCallSpy = jest.spyOn(makeApiCall, 'default').mockImplementation(async () => mockDB.cart.Data);
  //   const getUserSpy = jest.spyOn(UserService, 'getUser').mockImplementation(async () => pivotAPIUser(mockUserWithPreferredStore));

  //   const oldCart = await getCart();
  //   const cartData = await addCartItem({
  //     action: 0,
  //     product: [{ skuId: 1, qty: 2, productType: 2 }]
  //   });
  //   expect(cartData).toMatchSnapshot();

  //   expect(makeApiCallSpy).toHaveBeenCalledTimes(3);
  //   expect(makeApiCallSpy).toHaveBeenNthCalledWith(2, apiRoutes.addCartItem, {
  //     body: {
  //       userId: mockUserWithPreferredStore.userId,
  //       cartId: oldCart.cartId,
  //       action: 0,
  //       product: [{ skuId: 1, qty: 2, productType: 2 }],
  //       storeId: mockUserWithPreferredStore.preferences.storeId
  //     },
  //     headers: { StoreId: `${mockUserWithPreferredStore.preferences.storeId}` }
  //   });

  //   makeApiCallSpy.mockRestore();
  //   getUserSpy.mockRestore();
  // });
});
