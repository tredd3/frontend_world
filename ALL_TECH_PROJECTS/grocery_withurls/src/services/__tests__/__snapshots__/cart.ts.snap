// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart service tests for user with no selected store in preferences getCart for user with no default store 1`] = `
Object {
  "cartId": 2764873131,
  "couponDiscountAmount": 20,
  "deliveryCharge": 0,
  "discountAmount": 1.8999999999999986,
  "itemCount": 3,
  "savedForLater": null,
  "shipments": Array [
    Object {
      "couponDiscountAmount": 20,
      "creationDate": "17-Apr-2020",
      "deliveryCharge": 0,
      "deliveryDate": "",
      "deliveryDay": "Tomorrow",
      "discountAmount": 1.8999999999999986,
      "itemCount": 3,
      "products": Array [
        Object {
          "categoryId": 10002340,
          "coupon": null,
          "id": 100024066,
          "image": "https://uat-grocery.ril.com:443//images/?name=images/hc1/hcb/8916991115294.jpg",
          "isFc": 1,
          "mrp": 240,
          "name": "Nivea Smooth Body Milk",
          "otherImages": Array [
            "https://uat-grocery.ril.com:443//images/?name=images/hc1/hcb/8916991115294.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h16/h3a/8916957954078.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hc8/h0a/8916962082846.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hf9/h66/8916966604830.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hcf/hfa/8916976173086.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h74/he3/8916983447582.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hf6/h4a/8916987183134.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h56/hce/8917003763742.jpg",
          ],
          "savingsPercent": 0,
          "selectedQuantity": 1,
          "skuId": 10028095,
          "sp": 240,
        },
        Object {
          "categoryId": 10004128,
          "coupon": Object {
            "code": "",
            "description": "http://www.google.com",
            "discountAmount": 10,
            "id": 4170,
            "title": "10 rs off on Catch Whole Jeera 100 g",
          },
          "id": 100008042,
          "image": "https://uat-grocery.ril.com:443//images/?name=images/h1c/h24/8917070282782.jpg",
          "isFc": 1,
          "mrp": 38,
          "name": "Catch Whole Jeera 100 g",
          "otherImages": Array [
            "https://uat-grocery.ril.com:443//images/?name=images/h1c/h24/8917070282782.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hc2/h06/8917068120094.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hef/h25/8917070151710.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h14/hee/8917072052254.jpg",
          ],
          "savingsPercent": 5,
          "selectedQuantity": 1,
          "skuId": 10009279,
          "sp": 36.1,
        },
        Object {
          "categoryId": 10003881,
          "coupon": Object {
            "code": "",
            "description": "Rs 10 off on Whiskas Ocean Fish Adult Cat Food 3kg",
            "discountAmount": 10,
            "id": 4009,
            "title": "Rs 10 off on Whiskas Ocean Fish Adult Cat Food 3kg",
          },
          "id": 100038166,
          "image": "https://uat-grocery.ril.com:443//images/?name=images/h64/he4/8916990001182.jpg",
          "isFc": 1,
          "mrp": 840,
          "name": "Whiskas Ocean Fish Adult Cat Food 3 kg",
          "otherImages": Array [
            "https://uat-grocery.ril.com:443//images/?name=images/h64/he4/8916990001182.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h37/hba/8917010710558.jpg",
          ],
          "savingsPercent": 0,
          "selectedQuantity": 1,
          "skuId": 10043518,
          "sp": 840,
        },
      ],
      "totalAmount": 1116.1,
      "totalPayable": 1096.1,
      "type": 1,
    },
  ],
  "totalAmount": 1116.1,
  "totalPayableAmount": 1096.1,
}
`;
